File: assets/Scripts/player/PlayerController.js
// FILE: Scripts/player/PlayerController.js
"use strict";

const { PlayerPawn } = require("./PlayerPawn.js");
const { createPlayerControllers } = require("./PlayerControllers.js");

function req(v, msg) {
    if (v == null) throw new Error(msg);
    return v;
}

/**
 * PlayerController
 *
 * Owns PlayerPawn + all controllers.
 * No "bind" magic: every controller receives the player instance in constructor.
 */
class PlayerController {
    constructor(ctx, cfg) {
        this.ctx = req(ctx, "[Player] ctx required");
        this.cfg = cfg || null;

        this.pawn = new PlayerPawn(this.ctx, this.cfg).init();

        // Provide explicit access point for everyone (no hidden globals).
        this.controllers = [];
        this._started = false;

        // Create controllers with strict OOP injection.
        this.controllers = createPlayerControllers(this);

        // Deterministic lifecycle.
        this.start();
    }

    start() {
        if (this._started) return;
        for (let i = 0; i < this.controllers.length; i++) {
            const c = this.controllers[i];
            if (c && typeof c.onStart === "function") c.onStart();
        }
        this._started = true;
    }

    update(dt) {
        if (!this._started) this.start();
        for (let i = 0; i < this.controllers.length; i++) {
            const c = this.controllers[i];
            if (c && typeof c.onUpdate === "function") c.onUpdate(dt);
        }
    }

    dispose() {
        const list = this.controllers;

        this.controllers = [];
        this._started = false;

        // Stop controllers in reverse order.
        for (let i = list.length - 1; i >= 0; i--) {
            const c = list[i];
            if (c && typeof c.onStop === "function") {
                try { c.onStop(); } catch (_) {}
            }
        }

        const pawn = this.pawn;
        this.pawn = null;

        const ctx = this.ctx;
        this.ctx = null;
        this.cfg = null;

        if (pawn && typeof pawn.destroy === "function") {
            try { pawn.destroy(); } catch (_) {}
        }

        // ctx is managed by world runtime; do not dispose it here.
        void ctx;
    }
}

module.exports = { PlayerController };

File: assets/Scripts/player/PlayerController.lua
local M = {}
local luaRuntime = require("@builtin/lua_runtime")
local Error = luaRuntime.Error
local LuaConstruct = luaRuntime.LuaConstruct
local LuaClass = luaRuntime.LuaClass
local lua_require_result_0 = require("./PlayerPawn.lua")
PlayerPawn = lua_require_result_0.PlayerPawn
function req(self, v, msg)
    if v == nil then
        error(
            LuaConstruct(Error, msg),
            0
        )
    end
    return v
end
PlayerController = LuaClass()
PlayerController.name = "PlayerController"
function PlayerController.prototype.lua_constructor(self, ctx, cfg)
    self.ctx = req(_G, ctx, "[PlayerController] ctx required")
    self.cfg = cfg or nil
    self.pawn = LuaConstruct(PlayerPawn, self.ctx, self.cfg):init()
    local ENGINE = req(_G, _G.ENGINE, "[PlayerController] ENGINE required")
    req(_G, ENGINE.controllers, "[PlayerController] ENGINE.controllers required")
    self.ec = ENGINE.controllers:entity("player", self.ctx, self.pawn, self.cfg)
    if not self.ec then
        error(
            LuaConstruct(Error, "[PlayerController] ENGINE.controllers.entity(...) returned null"),
            0
        )
    end
end
function PlayerController.prototype.update(self, dt)
    local ec = self.ec
    if ec then
        ec:update(dt)
    end
end
function PlayerController.prototype.dispose(self)
    local ec = self.ec
    local pawn = self.pawn
    self.ec = nil
    self.pawn = nil
    self.ctx = nil
    self.cfg = nil
    if ec and KTypeOf(ec.dispose) == "function" then
        do
            pcall(function()
                ec:dispose()
            end)
        end
    end
    if pawn and KTypeOf(pawn.destroy) == "function" then
        do
            pcall(function()
                pawn:destroy()
            end)
        end
    end
end
M = {PlayerController = PlayerController}

return M
